src/shared/lock-util.c

```diff
@@ -115,3 +115,35 @@ void release_lock_file(LockFile *f) {
         f->fd = safe_close(f->fd);
         f->operation = 0;
 }
+
+int lockf_unposix(int fd, int cmd, off_t len) {
+        struct flock fl = {
+                .l_whence = SEEK_CUR,
+                .l_len = len
+        };
+
+        /* A version of lockf() that uses open file description locks instead of regular POSIX locks. OFD
+         * locks are per file descriptor instead of process wide. This function doesn't support F_TEST for
+         * now until we have a use case for it somewhere. */
+
+        assert(fd >= 0);
+
+        switch (cmd) {
+        case F_ULOCK:
+                fl.l_type = F_UNLCK;
+                cmd = F_OFD_SETLK;
+                break;
+        case F_LOCK:
+                fl.l_type = F_WRLCK;
+                cmd = F_OFD_SETLKW;
+                break;
+        case F_TLOCK:
+                fl.l_type = F_WRLCK;
+                cmd = F_OFD_SETLK;
+                break;
+        default:
+                return -EINVAL;
+        }
+
+        return RET_NERRNO(fcntl(fd, cmd, &fl));
+}
```

Comment:
btw, are we sure we want this wrapper like this? I think the fact it doesn't allow offsets to be specified and operates on the seek pointer is really a bad design decision. And the fact it allows negative offsets is also bad...

i am not convinced lockf() is really a great inspiration for anything, maybe let's just add a wrapper around unposix locks and use the command names they use? i.e. a simple:
```c

int unposix_lock_full(int fd, int cmd, int type, uint64_t start, uint64_t length) {
        if (start > INT64_MAX)
                return -EINVAL;
        if (length == 0) 
                return -EINVAL;
        if (length == UINT64_MAX)
                length = 0; /* fcntl() uses 0 for 'whole range' */
        else if (length > INT64 - start)
                return -EINVAL;

        return RET_NERRNO(fcntl(fd, cmd, &(struct flock) {
                        .l_type = type,
                        .l_whence = SEEK_SET,
                        .l_start = start,
                        .l_len = length,
        };
}

static inline unposix_lock(int fd, int cmd, int type) {
        return unposix_lock_ful(fd, cmd, type, 0, UINT64_MAX);
}
```
And in most cases (all?) we'd probably use the latter wrapper. (If we never need offset/size locks, then maybe even drop the full version?)

Reply:
I ended up implementing just unposix_lock() and giving it the flock() interface so we can use the same constants throughout the tree regardless of which type of lock we're using.

Reply:
yeah, the BSD flock() API is really nice and simple. I like.